engine/src/conversion/analysis/fun/implicit_constructors.rs

```diff
@@ -59,254 +75,441 @@ struct ExplicitFound {
-                .iter()
-                .chain(field_types.iter())
-                .any(|qn| !all_known_types.contains(qn));
-            let explicit_items_found = ExplicitItemsFound {
-                move_constructor: find(ExplicitKind::MoveConstructor),
-                copy_constructor: find(ExplicitKind::ConstCopyConstructor)
-                    || find(ExplicitKind::NonConstCopyConstructor)
-                    || find(ExplicitKind::DeletedOrInaccessibleCopyConstructor),
-                any_other_constructor: find(ExplicitKind::OtherConstructor),
-                any_bases_or_fields_lack_const_copy_constructors,
-                any_bases_or_fields_have_deleted_or_inaccessible_copy_constructors,
-                any_bases_have_deleted_or_inaccessible_destructors,
-                destructor: find(ExplicitKind::Destructor)
-                    || find(ExplicitKind::DeletedOrInaccessibleDestructor),
-                copy_assignment_operator: find(ExplicitKind::CopyAssignmentOperator),
-                move_assignment_operator: find(ExplicitKind::MoveAssignmentOperator),
-                has_rvalue_reference_fields: details.has_rvalue_reference_fields,
-                any_field_or_base_not_understood,
+            let get_items_found = |qn: &QualifiedName| -> Option<ItemsFound> {
+                if known_types().is_known(qn) {
+                    Some(known_type_items_found(qn))
+                } else {
+                    all_items_found.get(qn).copied()
+                }
             };
-            log::info!(
-                "Explicit items found for {:?}: {:?}",
-                name,
-                explicit_items_found
-            );
-            let implicits = determine_implicit_constructors(explicit_items_found);
-            implicit_constructors_needed.insert(name.clone(), implicits);
+            let bases_items_found: Vec<_> = bases.iter().map_while(get_items_found).collect();
+            let field_types_items_found: Vec<_> =
+                field_types.iter().map_while(get_items_found).collect();
+            let has_rvalue_reference_fields = details.has_rvalue_reference_fields;
+
+            // Check that all the bases and field types are known first. This combined with
+            // iterating via [`depth_first`] means we can safely search in `items_found` for all of
+            // them.
+            //
+            // Conservatively, we will not acknowledge the existence of defaulted special member
+            // functions for any struct/class where we don't fully understand all field types.
+            // However, we can still look for explictly declared versions and use those.
+            //
+            // We need to extend our knowledge to understand the constructor behavior of things in
+            // known_types.rs, then we'll be able to cope with types which contain strings,
+            // unique_ptrs etc.
+            //
+            // We also need to learn to follow typedefs.
```

Comment:
This is the only part I'm worried about breaking previously-working setups with what I consider "reasonable C++". For example, a few of the tests needed workarounds for this. I couldn't figure out how to resolve the final targets of typedefs at this stage though.

Reply:
Typedefs are actually fine, I was confused.